File: 2021-06-June-LeetCoding-Challenge/Longest Consecutive Sequence.py
# -*- coding: utf-8 -*-
from typing import List


class Solution:
    '''
    Top-down dynamic programming approach.
    '''
    def longestConsecutive(self, nums: List[int]) -> int:
        nums = set(nums)  # Duplicate numbers need not be computed
        max_count = int()
        cache = dict()

        def dp(num):
            if num in cache:
                # Consecutive count for number already computed
                return cache.get(num)

            count = 1
            if num-1 in nums:
                # If num-1 is also present, compute num-1
                count += dp(num-1)

            cache[num] = count  # Store num's consecutive count in cache
            return count

        for num in nums:
            max_count = max(max_count, dp(num))
        return max_count


if __name__ == '__main__':
    obj = Solution()
    nums = [100, 4, 200, 1, 3, 2]
    assert obj.longestConsecutive(nums) == 4

    obj = Solution()
    nums = [0, 3, 7, 2, 5, 8, 4, 6, 0, 1]
    assert obj.longestConsecutive(nums) == 9

File: 2022-07-July-LeetCoding-Challenges/Longest Consecutive Sequence.go
package main

func max(n1, n2 int) int {
    if n1 >= n2 {
        return n1
    }
    return n2
}

func longestConsecutive(nums []int) (maxCount int) {
    numSet := make(map[int]bool)
    // create set of nums
    for _, num := range nums {
        numSet[num] = true
    }

    for num, _ := range numSet {
        count := 1
        // if next lower is present,
        // consecutive is already checked
        if numSet[num-1] {
            continue
        }
        for {
            // check for next consecutive
            if numSet[num+1] {
                num++
                count++
            } else {
                break
            }
        }
        maxCount = max(maxCount, count)
    }
    return maxCount
}

func main() {
    nums := []int{0,3,7,2,5,8,4,6,0,1}
    expected := 9
    actual := longestConsecutive(nums)
    if actual != expected {
        panic("actual differs from expected")
    }
}
